pkg/apis/alicloud/validation/shoot.go

```diff
@@ -19,21 +19,33 @@ import (
 	"regexp"
 
 	apisalicloud "github.com/gardener/gardener-extension-provider-alicloud/pkg/apis/alicloud"
-
 	"github.com/gardener/gardener/pkg/apis/core"
 	"github.com/gardener/gardener/pkg/apis/core/validation"
+	cidrvalidation "github.com/gardener/gardener/pkg/utils/validation/cidr"
 	apivalidation "k8s.io/apimachinery/pkg/api/validation"
 	"k8s.io/apimachinery/pkg/util/sets"
 	utilvalidation "k8s.io/apimachinery/pkg/util/validation"
 	"k8s.io/apimachinery/pkg/util/validation/field"
 )
 
-// ValidateNetworking validates the network settings of a Shoot.
+// AliCloudVPCCidr is a IPV4 CIDR used within AliCloud.
+// For example: the meta service endpoint is 100.100.100.200.
+// This CIDR can be accessed by any machine which is running with AliCloud VPC.
+const AliCloudVPCCidr = "100.64.0.0/10"
+
+// ValidateNetworking validates the network settings of a Shoot during creation.
+// As the network settings are immutable, we don't need to do validation for update.
 func ValidateNetworking(networking core.Networking, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
 
-	if networking.Nodes == nil {
-		allErrs = append(allErrs, field.Required(fldPath.Child("nodes"), "a nodes CIDR must be provided for Alicloud shoots"))
+	if networking.Nodes != nil && cidrvalidation.NetworksIntersect(*networking.Nodes, AliCloudVPCCidr) {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("nodes"), *networking.Nodes, "must not overlap with 100.64.0.0/10 because 10.64.0.0/10 is reserved by AliCloud"))
+	}
+	if networking.Pods != nil && cidrvalidation.NetworksIntersect(*networking.Pods, AliCloudVPCCidr) {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("pods"), *networking.Pods, "must not overlap with 100.64.0.0/10 because 10.64.0.0/10 is reserved by AliCloud"))
+	}
+	if networking.Services != nil && cidrvalidation.NetworksIntersect(*networking.Services, AliCloudVPCCidr) {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("services"), *networking.Services, "must not overlap with 100.64.0.0/10 because 10.64.0.0/10 is reserved by AliCloud"))
```

Comment:
All these fields (nodes/pods/services) are required for Alicloud. So you should enforce that these fields are not nil. Currently, you only check for intersection if they are not nil, but this would allow end-users to create a shoot without the nodes CIDR, for example.

Reply:
pods/services cidrs are ensured not nil in Gardener core service: https://github.com/gardener/gardener/blob/master/pkg/utils/validation/cidr/disjoint.go#L37
Regarding node cidrs, you are right. Still need to check nil.

Reply:
Yes, but this validation will probably go away soon. Actually, that's a good example why any external code should not rely on what Gardener validates and what not - it is not subject to any API contract or reliable in any other way. This kind of invalidates what we discussed earlier, i.e., also on shoot updates the CIDR immutability should be checked.

Reply:
But from my perspective, I would like to check kubernetes network config (nodes/pods/service) in the core because these fields are exposed in Gardener core API. The extenstion validator only validates those specialties for cloudprovider specific.

Reply:
If such kind of validation will not happen in in-tree admission, shall we provide a generic check in common framwork of extension?

Reply:
> because these fields are exposed in Gardener core API.

Yes, those are fields in the core, but it is infrastructure-specific whether they are required or not. For most hyperscalers they are required, but in private clouds or custom infrastructures not necessarily. Hence, the validation should happen here to ensure that ALL the fields needed by Alicloud extension are present.

> If such kind of validation will not happen in in-tree admission, shall we provide a generic check in common framwork of extension?

We could think of something like this in the future, but for now it's straight-forward to simply do it here (without going the de-tour of the extension library, waiting for a new Gardener release, and then call the introduced function).